py/dml/ctree.py

```diff
@@ -3342,9 +3350,9 @@ def ctype(self):
         return TPtr(self.type_)
 
     def read(self):
-        return '(%s)((uintptr_t)_dev + %s->%s)' % (
+        return '((%s)((uintptr_t)_dev + %s->%s) + (%s).id.encoded_index)' % (
```

Comment:
This duplicates side-effects. You need to turn this into a statement expression with an intermediary variable.

Reply:
Good catch. Probably deserves a test, too.